pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/parser/Converter.groovy

```diff
@@ -201,12 +202,14 @@ public class Converter {
      * @param run The {@link WorkflowRun} to pull from.
      * @return A parsed and validated {@link ModelASTPipelineDef}
      */
-    public static ModelASTPipelineDef parseFromWorkflowRun(WorkflowRun run) {
-        CpsFlowExecution execution = Utils.getExecutionForRun(run)
-
-        String rawScript = execution.script
-
-        return scriptToPipelineDef(rawScript)
+    public static ModelASTPipelineDef parseFromWorkflowRun(WorkflowRun run) throws Exception {
+        CpsFlowExecution execution = null
+        if (run.execution != null) {
+            execution = (CpsFlowExecution) run.execution
+        } else if (run.getExecutionPromise() != null) {
+            execution = (CpsFlowExecution) run.getExecutionPromise().get(2, TimeUnit.SECONDS)
```

Comment:
So I need the parsing to be synchronous, meaning I don't *think* I can do this in a callback? I could be wrong - I don't entirely grok callbacks.